label_studio/io_storages/proxy_api.py

```diff
@@ -0,0 +1,327 @@
+        4 'bytes=-1024': Handles negative start (not supported)
+        
+        Returns:
+            str: Modified range header in format "bytes=start-end" or None if no range header
+        """
+        max_range_size = settings.RESOLVER_PROXY_MAX_RANGE_SIZE
+        range_header = None
+
+        if rng := request.headers.get('Range'):
+            start, end = 0, ''
+            try:
+                values = rng.split("=")[1].split("-")
+                start = int(values[0])
+                if len(values) > 1:
+                    end = values[1]
+                    if end != '':
+                        end = int(end)
+                logger.debug(f">> range read from request: start: {start}, end: {end}")
+            except (IndexError, ValueError) as e:
+                logger.warning(f"Invalid range header: {rng}: {e}")
+                start = 0
+                end = ''
+
+            """
+            Pass this range as is to storage:
+              - 'bytes=0-'  most likely, browser is requesting just headers
+              - 'bytes=0-0'  most likely, browser is requesting just headers
+            Limit stream to MAX_RANGE bytes:
+              - 'bytes=123456-'  browser is requesting from 123456 to the end of the file
+              - 'bytes=123456-0'  browser is requesting from 123456 to the end of the file
+              - 'bytes=123456-789012'  browser is requesting from 123456 to 789012
+            Not supported:
+              - 'bytes=-1024'  browser is requesting last 1024 bytes - we don't support this
+            """
+            # 'bytes=0-' + 'bytes=0-0'
+            if start == 0 and (end == '' or end == 0):
+                pass
+            # 'bytes=123456-' + 'bytes=123456-0'
+            elif start > 0 and (end == '' or end == 0):
+                end = start + max_range_size
+            # 'bytes=123456-' + 'bytes=123456-789012'
+            elif start >= 0 and end > 0:
+                end = start + max_range_size if end >= start + max_range_size else end
+            # 'bytes=-1024'
+            elif start < 0:
+                logger.warning(f"Start range is negative and not supported: {rng}")
+                start = 0
+                end = start + max_range_size
+            else:
+                logger.warning(f"Range is not covered by logic: {rng}")
+                start = 0
+                end = ''
+
+            range_header = f"bytes={start}-{end}"
+            logger.debug(f">> stream > start: {int(start)/1024/1024} MB")
+            logger.debug(f">> stream > end: {int(end or 0)/1024/1024} MB")
+            logger.debug(f">> stream > range_header: {range_header}")
+            
+        return range_header
+
+    def proxy_data_from_storage(self, request, uri, project, storage):
+        """
+        Proxy the data using iter_chunks directly from storage streaming object.
+        
+        This implementation forwards Range headers to S3 and streams the response
```

Comment:
same question about S3 here, is this S3 specific?

Reply:
fixed. it's for any storages